source/CCustomOpcodeSystem.cpp

```diff
@@ -265,6 +269,123 @@ namespace CLEO
+			return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+		}
+
+		int nVarArg = GetVarArgCount(thread);
+		if (numParams + returnArg != nVarArg) // and return argument
+		{
+			SHOW_ERROR("Opcode [%04X] declared %d input args, but provided %d in script %s\nScript suspended.", opcode, numParams, (int)nVarArg - returnArg, ((CCustomScript*)thread)->GetInfoStr().c_str());
+			return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+		}
+
+		constexpr size_t Max_Args = 32;
+		if (numParams > Max_Args)
+		{
+			SHOW_ERROR("Opcode [%04X] used with more than supported arguments in script %s\nScript suspended.", opcode, ((CCustomScript*)thread)->GetInfoStr().c_str());
+			return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+		}
+
+		static SCRIPT_VAR arguments[Max_Args] = { 0 };
+		SCRIPT_VAR* arguments_end = arguments + numParams;
+
+		constexpr size_t Max_Text_Params = 5;
+		static char textParams[Max_Text_Params][MAX_STR_LEN];
+		size_t currTextParam = 0;
+
+		stackAlign *= 4; // bytes peer argument
+		
+		// retrieve parameters
+		for (size_t i = 0; i < numParams; i++)
+		{
+			auto& param = arguments[i];
+			auto paramType = thread->PeekDataType();
+
+			if (IsImmInteger(paramType) || IsVariable(paramType))
+				*thread >> param.dwParam;
+			else
+			if (IsImmFloat(paramType))
+				*thread >> param.fParam;
+			else
+			if (IsImmString(paramType) || IsVarString(paramType))
+			{
+				if (currTextParam >= Max_Text_Params)
+				{
+					SHOW_ERROR("Opcode [%04X] used with more than supported string arguments in script %s\nScript suspended.", opcode, ((CCustomScript*)thread)->GetInfoStr().c_str());
+					return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+				}
+
+				param.pcParam = ReadStringParam(thread, textParams[currTextParam], MAX_STR_LEN); OPCODE_VALIDATE_STR_ARG_READ(param.pcParam)
+				currTextParam++;
+			}
+			else
+			{
+				SHOW_ERROR("Invalid param type (%s) in opcode [%04X] in script %s \nScript suspended.", opcode, ToKindStr(paramType), ((CCustomScript*)thread)->GetInfoStr().c_str());
+				return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+			}
+		}
+
+		// validate return target variable
+		if (returnArg)
+		{
+			auto paramType = thread->PeekDataType();
+
+			if (!IsVariable(paramType) && !IsVarString(paramType))
+			{
+				SHOW_ERROR("Invalid return param type (%s) in opcode [%04X] in script %s \nScript suspended.", opcode, ToKindStr(paramType), ((CCustomScript*)thread)->GetInfoStr().c_str());
+				return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+			}
+		}
+
+		DWORD result;
+		_asm
+		{
+			// transfer args to stack
+			lea ecx, arguments
+			call_func_loop:
+			cmp ecx, arguments_end
+				jae call_func_loop_end
+				push[ecx]
+				add ecx, 0x4
+				jmp call_func_loop
+				call_func_loop_end:
+
+			// set struct pointer for this call
+			lea ecx, thisCall
+			test ecx, ecx
+			jz call_func_this_end
+			mov ecx, struc
+			call_func_this_end:

```

Comment:
micro-optimization:

since you use ecx for intermediate calculation I don't see much harm in assigning `ecx` to `struct` regardless of `thisCall` value. It could be a garbage value, could be not, the callee will use it if needed.

ie.

```
// set struct pointer for this call
mov ecx, struc
// call function
xor eax, eax
call func
```

Reply:
True. Even better, as unused `struc` has value 0.
Updated.